File: src/com/gmail/marcosav2010/communicator/packet/Packet.java
package com.gmail.marcosav2010.communicator.packet;

import java.io.IOException;

import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder;
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder;

import lombok.Getter;

public abstract class Packet extends AbstractPacket {

	public static int MAX_SIZE = AbstractPacket.MAX_SIZE - Integer.BYTES;

	@Getter
	private int packetID = 0;

	public Packet setPacketID(int packetID) {
		if (hasPacketID())
			throw new RuntimeException("This packet already has an ID set");

		this.packetID = packetID;
		return this;
	}

	private boolean hasPacketID() {
		return packetID != 0;
	}

	public boolean shouldSendRespose() {
		return true;
	}

	@Override
	public boolean isStandard() {
		return false;
	}

	@Override
	public void encode(PacketEncoder out) throws IOException {
		if (!hasPacketID())
			throw new RuntimeException("Cannot encode without ID set");

		out.write(packetID);
		encodeContent(out);
	}

	protected abstract void encodeContent(PacketEncoder out) throws IOException;

	@Override
	public void decode(PacketDecoder in) throws IOException {
		packetID = in.readInt();
		decodeContent(in);
	}

	protected abstract void decodeContent(PacketDecoder in) throws IOException;
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/Packet.kt
package com.gmail.marcosav2010.communicator.packet

import com.gmail.marcosav2010.communicator.packet.wrapper.PacketDecoder
import com.gmail.marcosav2010.communicator.packet.wrapper.PacketEncoder

abstract class Packet : AbstractPacket() {

    companion object {
        var MAX_SIZE = AbstractPacket.MAX_SIZE - Integer.BYTES
    }

    var packetID = 0
        private set

    fun setPacketID(packetID: Int): Packet {
        if (hasPacketID()) throw RuntimeException("This packet already has an ID set")
        this.packetID = packetID
        return this
    }

    private fun hasPacketID() = packetID != 0

    open fun shouldSendResponse() = true

    override val isStandard
        get() = false

    override fun encode(out: PacketEncoder) {
        if (!hasPacketID()) throw RuntimeException("Cannot encode without ID set")
        out.write(packetID)
        encodeContent(out)
    }

    protected abstract fun encodeContent(out: PacketEncoder)

    override fun decode(`in`: PacketDecoder) {
        packetID = `in`.readInt()
        decodeContent(`in`)
    }

    protected abstract fun decodeContent(`in`: PacketDecoder)
}